packages/cw721/src/execute.rs

```diff
@@ -0,0 +1,810 @@
+
+    // update contract version
+    cw2::set_contract_version(storage, contradct_name, contract_version)?;
+    Ok(response)
+}
+
+pub fn migrate_creator(
+    storage: &mut dyn Storage,
+    api: &dyn Api,
+    _env: &Env,
+    msg: &Cw721MigrateMsg,
+    response: Response,
+) -> StdResult<Response> {
+    match msg {
+        Cw721MigrateMsg::WithUpdate { creator, .. } => {
+            if let Some(creator) = creator {
+                CREATOR.initialize_owner(storage, api, Some(creator.as_str()))?;
+                return Ok(response.add_attribute("creator", creator));
+            }
+        }
+    }
+    Ok(response)
+}
+
+pub fn migrate_minter(
+    storage: &mut dyn Storage,
+    api: &dyn Api,
+    _env: &Env,
+    msg: &Cw721MigrateMsg,
+    response: Response,
+) -> StdResult<Response> {
+    match msg {
+        Cw721MigrateMsg::WithUpdate { minter, .. } => {
+            if let Some(minter) = minter {
+                MINTER.initialize_owner(storage, api, Some(minter.as_str()))?;
+                return Ok(response.add_attribute("creator", minter));
+            }
+        }
+    }
+    Ok(response)
+}
+
+/// Migrates only in case ownership is not present
+/// !!! Important note here: !!!
+/// - creator owns the contract and can update collection info
+/// - minter can mint new tokens
+///
+/// Before v0.19.0 there were confusing naming conventions:
+/// - v0.17.0: minter was replaced by cw_ownable, as a result minter is owner
+/// - v0.16.0 and below: minter was stored in dedicated `minter` store (so NOT using cw_ownable at all)
+pub fn migrate_legacy_minter_and_creator(
+    storage: &mut dyn Storage,
+    api: &dyn Api,
+    _env: &Env,
+    _msg: &Cw721MigrateMsg,
+    response: Response,
+) -> Result<Response, Cw721ContractError> {
+    let minter = MINTER.item.may_load(storage)?;
+    // no migration in case minter is already set
+    if minter.is_some() {
+        return Ok(response);
+    }
+    // in v0.17/18 cw_ownable::OWNERSHIP was used for minter, now it is used for creator
+    let ownership_previously_used_as_minter = CREATOR.item.may_load(storage)?;
+    let creator_and_minter = match ownership_previously_used_as_minter {
+        // v0.18 migration
+        Some(ownership) => {
+            // owner is used for both: creator and minter
+            // since it is already set for creator, we only need to migrate minter
+            let owner = ownership.owner.map(|a| a.to_string());
+            MINTER.initialize_owner(storage, api, owner.as_deref())?;
+            owner
+        }
+        // v0.17 and older migration
+        None => {
+            let legacy_minter_store: Item<Addr> = Item::new("minter");
+            let legacy_minter = legacy_minter_store.load(storage)?;
+            MINTER.initialize_owner(storage, api, Some(legacy_minter.as_str()))?;
+            CREATOR.initialize_owner(storage, api, Some(legacy_minter.as_str()))?;
+            Some(legacy_minter.to_string())
+        }
+    };
```

Comment:
...then we have 2 legacy cases:
1. v0.17/18 where minter is stored in `cw-ownable`
2. v0.16 and older where minter is stored in dedicated `minter: Item` store.

In case of 1) there is `Some(ownership)` and this is used for new MINTER ownership.

in case of 2) there is none in ownership store, but in very old minter/item store.